### DISCUSSION TITLE
Do tokio reuse spawn_blocking threads?

### DISCUSSION THREAD

**Discussion by ghost** [tokio-rs/tokio#3251] [Q&A] [answered]
Hi,

I was wondering what is best to execute blocking tasks for a web  service. So I was asking myself: Do tokio reuse threads from the blocking threadpool launched with `spawn_blocking`?

from my basic tests it seems so. When calling 100 times  a function which launch 2 `spawn_blocking`:

```rust
pub async fn route_blocking() -> HttpResponse {
    let _ = tokio::task::spawn_blocking(|| {
        let duration = std::time::Duration::from_secs(2);
        std::thread::sleep(duration);
    }).await;
    let res = tokio::task::spawn_blocking(|| {
        let duration = std::time::Duration::from_secs(2);
        std::thread::sleep(duration);
        let mut data = HashMap::new();
    data.insert("hello", "from blocking2");
    Response::data(data)
    }).await.unwrap();


    HttpResponse::Ok().json(res)
}
```

```sh
# 100 concurrent calls
$ ab -c 100 -n 500 http://127.0.0.1:8888/api/blocking
```

my process creates only ~120 threads which may indicate that the threads a reused.

Or should I use a third-party, fixed-size threadpool like `rayon`?

**Comment by hawkw** [comment, marked as answer]
Tokio [uses a dedicated thread pool of blocking worker threads](https://docs.rs/tokio/0.3.5/tokio/task/index.html#spawn_blocking) for running blocking tasks. When a blocking worker completes its current blocking task, it returns to the blocking pool and will execute new blocking work when new blocking tasks are spawned.

When all blocking workers in the pool are busy, new blocking threads will be spawned, up to a limit configured by the [`max_threads`](https://docs.rs/tokio/0.3.5/tokio/runtime/struct.Builder.html#method.max_threads) runtime builder setting. Skimming the docs, it seems like this might not be stated clearly enough in the API documentation, we could definitely add something to clarify this.

**Comment by ghost** [comment]
Crystal clear, thank you again @hawkw !

I also noticed that after a 'short' amount of time without requests, the threads count of my application go down from ~120 to ~6. As if `tokio` killed threads of my blocking threadpool.

Blocking threads are killed after `KEEP_ALIVE` time right? 
https://github.com/tokio-rs/tokio/blob/52cd240053b2e1dd5835186539f563c3496dfd7d/tokio/src/runtime/blocking/pool.rs#L76

If it's true, may I try a small PR to make it more clear in the docs tomorrow?

**Reply by Darksonn** [reply]
PRs are always welcome!

**Reply by Darksonn** [reply]
You can configure the keep-alive with https://docs.rs/tokio/0.3/tokio/runtime/struct.Builder.html#method.thread_keep_alive

**Reply by ghost** [reply]
https://github.com/tokio-rs/tokio/pull/3253
